### PR TITLE
Deploy new API rr_node.node_sides() to GUI

### PR DESCRIPTION
**Pull request by tangxifan**
This is a follow up pull request for the PR #1621 
The purpose of this PR is to deploy the new APIs about the node sides of RRGraph to the *Graph User Interface* (GUI)

**Note**
This PR will not deprecate the old APIs and the unnecessary node_sides() APIs.
The deprecation will be done after lookahead map has been fixed, as the final PR in this series of patches

#### Description

Major changes are landed in the GUI part 
In the GUI part, the major changes to enable the drawing for lines between the IPIN/OPIN and CHANX/CHANY routing resource nodes.
As an IPIN/OPIN may have multiple sides, the drawer should identify the correct line (from the multiple possible connections) to display.

** A Side Note**
A separated PR will be created to solve the issue on the router lookahead map (as stated in issue #1580)
In the router lookahead map, a proper side should be picked from multiple sides of a rr_node so that the lookahead map can have an accurate estimation on the routing congestion.
As multiple sides are used in Titan architectures, a proof of the QoR changes is required!

#### Related Issue

To address issue #1577 

#### Motivation and Context

To show correct RRGraph display for nodes/edges with multiple sides in GUI.

#### How Has This Been Tested?

**As multiple sides are used in Titan architectures, a proof of the GUI display is required!**

My major focus is on the Titan architecture where multiple sides are heavily used for rr_nodes.
- Before GUI patching (LAB in Titan Architecture): 
image

- After GUI patching (LAB in Titan Architecture): You can see the connections on the top side has been correctly shown, which are driven by LAB.data_in (locate on TOP, LEFT, RIGHT sides)
image

- Before GUI patching (PLL and I/Os in Titan Architecture): 
image

- After GUI patching (PLL and I/Os in Titan Architecture): 
image

- Before GUI patching (M9K in Titan Architecture): 
image

- After GUI patching (M9K in Titan Architecture):
image

- Before GUI patching (DSP36 in Titan Architecture): 
image

- After GUI patching (DSP36 in Titan Architecture): 
image

**Note that there is no difference on normal architectures.
- Before GUI patching (k6_N10_mem32K_40nm.xml): 
image

- After GUI patching (k6_N10_mem32K_40nm.xml): 
image

The changes will only affect GUI codes under ``vpr/src/draw``.

#### Types of changes

- [X] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by tangxifan** [comment]
@vaughnbetz @litghost I have finished my patch on the GUI using the new node_side() APIs.
All the graphic proofs have been uploaded. This PR is ready for your review.

File: vpr/src/draw/draw.cpp [hunk 1]
```diff
@@ -2955,19 +2963,83 @@ static void draw_pin_to_chan_edge(int pin_node, int chan_node, ezgl::renderer* g
 
     const t_grid_tile& grid_tile = device_ctx.grid[pin_rr.xlow()][pin_rr.ylow()];
     t_physical_tile_type_ptr grid_type = grid_tile.type;
-    VTR_ASSERT_MSG(grid_type->pinloc[grid_tile.width_offset][grid_tile.height_offset][pin_rr.side()][pin_rr.pin_num()],
-                   "Pin coordinates should match block type pin locations");
 
     float draw_pin_offset;
-    if (pin_rr.side() == TOP || pin_rr.side() == RIGHT) {
```
Comment (litghost) [inline vpr/src/draw/draw.cpp:2962]: I think this logic needs a comment.  Now that pin rr nodes have multiple edges, I don't think this logic is right.  I recommend you write a short comment explaining why you believe this is correct.

Comment (tangxifan) [reply vpr/src/draw/draw.cpp:2962]: Indeed. In the later part of this code block, I have logics to identify which side to be used for GUI. Therefore, I move the offset calculation there.

**Review by litghost — APPROVED** [review]
LGTM

**Review by vaughnbetz — COMMENTED** [review]
LGTM.  Only had one suggestion for a comment (same place Keith flagged).

File: vpr/src/draw/draw.cpp (hunk 1 above)
Comment (tangxifan) [reply vpr/src/draw/draw.cpp:2962]: @vaughnbetz No problem. I have added comments about why the positive/negative offset is needed

**Comment by vaughnbetz** [comment]
This is all graphics, so it won't affect the nightly tests.  Will merge as soon as sanitized complete, although I think this code is not active in them either.

**Pull request merged**